feeder/cmd/rekor_feeder/main.go

```diff
@@ -0,0 +1,195 @@
+	wit_http "github.com/google/trillian-examples/witness/golang/client/http"
+	yaml "gopkg.in/yaml.v2"
+)
+
+var (
+	configFile = flag.String("config_file", "", "Path to feeder config file.")
+	timeout    = flag.Duration("timeout", 10*time.Second, "Maximum time to wait for witnesses to respond.")
+	interval   = flag.Duration("interval", time.Duration(0), "Interval between attempts to feed checkpoints. Default of 0 causes the tool to be a one-shot.")
+)
+
+// Config encapsulates the feeder config.
+type Config struct {
+	// Logs defines the source logs to feed from.
+	Logs []config.Log `yaml:"Logs"`
+
+	// Witness is the configured witness.
+	Witness config.Witness `yaml:"Witness"`
+}
+
+func main() {
+	flag.Parse()
+
+	cfg, err := readConfig(*configFile)
+	if err != nil {
+		glog.Exitf("Failed to read config: %v", err)
+	}
+
+	u, err := url.Parse(cfg.Witness.URL)
+	if err != nil {
+		glog.Exitf("Failed to parse witness URL %q: %v", cfg.Witness.URL, err)
+	}
+	witness := wit_http.Witness{
+		URL:      u,
+		Verifier: mustCreateVerifier(i_note.Note, cfg.Witness.PublicKey),
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, l := range cfg.Logs {
+		wg.Add(1)
+		go func(l config.Log, w wit_http.Witness) {
+			defer wg.Done()
+			if err := feedLog(l, witness, *timeout, *interval); err != nil {
+				glog.Errorf("feedLog: %v", err)
+			}
+		}(l, witness)
+	}
+	wg.Wait()
+}
+
+// logInfo is a partial representation of the JSON object returned by Rekór's
+// api/v1/log request.
+type logInfo struct {
+	// SignedTreeHead contains a Rekór checkpoint.
+	SignedTreeHead string `json:"signedTreeHead"`
+}
+
+// proof is a partial representation of the JSON struct returned by the Rekór
+// api/v1/log/proof request.
+type proof struct {
+	Hashes [][]byte `json:"hashes"`
+}
+
+func feedLog(l config.Log, w wit_http.Witness, timeout time.Duration, interval time.Duration) error {
+	lURL, err := url.Parse(l.URL)
+	if err != nil {
+		return fmt.Errorf("invalid LogURL %q: %v", l.URL, err)
+	}
+
+	fetchCP := func(ctx context.Context) ([]byte, error) {
+		li := logInfo{}
+		if err := getJSON(ctx, lURL, "api/v1/log", &li); err != nil {
+			return nil, fmt.Errorf("failed to fetch log info: %v", err)
+		}
+		return []byte(li.SignedTreeHead), nil
+	}
+	fetchProof := func(ctx context.Context, from, to log.Checkpoint) ([][]byte, error) {
+		if from.Size == 0 {
+			return [][]byte{}, nil
+		}
+		cp := proof{}
+		if err := getJSON(ctx, lURL, fmt.Sprintf("api/v1/log/proof?firstSize=%d&lastSize=%d", from.Size, to.Size), &cp); err != nil {
+			return nil, fmt.Errorf("failed to fetch log info: %v", err)
+		}
+		return cp.Hashes, nil
+	}
+
+	opts := feeder.FeedOpts{
+		LogID:           l.ID,
+		LogOrigin:       l.Origin,
+		FetchCheckpoint: fetchCP,
+		FetchProof:      fetchProof,
+		LogSigVerifier:  mustCreateVerifier(l.PublicKeyType, l.PublicKey),
+		Witness:         w,
+	}
+	if interval > 0 {
+		return feeder.Run(context.Background(), interval, opts)
```

Comment:
Consider passing the context into this method (`feedLog`).

Reply:
done